subgraphs/_reference_/src/sdk/protocols/generic/protocolSnapshot.ts

```diff
@@ -0,0 +1,211 @@
+
+    snapshot.save();
+  }
+
+  private takeUsageDailySnapshot(day: i32): void {
+    const activity = this.activityHelper;
+
+    const snapshot = new UsageMetricsDailySnapshot(Bytes.fromI32(day));
+    const previousSnapshot = UsageMetricsDailySnapshot.load(
+      this.protocol.lastSnapshotDayID
+    );
+
+    snapshot.protocol = this.protocol.id;
+    snapshot.day = day;
+    snapshot.blockNumber = this.event.block.number;
+    snapshot.timestamp = this.event.block.timestamp;
+
+    // unique users
+    snapshot.cumulativeUniqueUsers = this.protocol.cumulativeUniqueUsers;
+    snapshot.cumulativeUniqueTransferSenders =
+      this.protocol.cumulativeUniqueTransferSenders;
+    snapshot.cumulativeUniqueTransferReceivers =
+      this.protocol.cumulativeUniqueTransferReceivers;
+    snapshot.cumulativeUniqueLiquidityProviders =
+      this.protocol.cumulativeUniqueLiquidityProviders;
+    snapshot.cumulativeUniqueMessageSenders =
+      this.protocol.cumulativeUniqueMessageSenders;
+
+    // daily activity
+    snapshot.dailyActiveUsers = activity.dailyActiveUsers;
+
+    // transaction counts
+    snapshot.cumulativeTransactionCount =
+      this.protocol.cumulativeTransactionCount;
+
+    // misc
+    snapshot.totalPoolCount = this.protocol.totalPoolCount;
+
+    // deltas
+    let transactionDelta = snapshot.cumulativeTransactionCount;
+
+    if (previousSnapshot) {
+      transactionDelta =
+        snapshot.cumulativeTransactionCount -
+        previousSnapshot.cumulativeTransactionCount;
+    }
+    snapshot.dailyTransactionCount = transactionDelta;
+    snapshot.save();
+
+    activity.dailyActiveUsers = 0;
+    activity.save();
+  }
+
+  private takeUsageHourlySnapshot(hour: i32): void {
+    const activity = this.activityHelper;
+
+    const snapshot = new UsageMetricsHourlySnapshot(Bytes.fromI32(hour));
+    const previousSnapshot = UsageMetricsHourlySnapshot.load(
+      this.protocol.lastSnapshotHourID
+    );
+
+    snapshot.protocol = this.protocol.id;
+    snapshot.hour = hour;
+    snapshot.blockNumber = this.event.block.number;
+    snapshot.timestamp = this.event.block.timestamp;
+
+    // unique users
+    snapshot.cumulativeUniqueUsers = this.protocol.cumulativeUniqueUsers;
+
+    // hourly activity
+    snapshot.hourlyActiveUsers = activity.hourlyActiveUsers;
+
+    // transaction counts
+    snapshot.cumulativeTransactionCount =
+      this.protocol.cumulativeTransactionCount;
+
+    // deltas
+    let transactionDelta = snapshot.cumulativeTransactionCount;
+    if (previousSnapshot) {
+      transactionDelta =
+        snapshot.cumulativeTransactionCount -
+        previousSnapshot.cumulativeTransactionCount;
+    }
+    snapshot.hourlyTransactionCount = transactionDelta;
+    snapshot.save();
+
+    activity.hourlyActiveUsers = 0;
+    activity.save();
+  }
+}
+
+function initActivityHelper(): _ActivityHelper {
+  let helper = _ActivityHelper.load(ActivityHelperID);
```

Comment:
You are missing this entity in the schema

Reply:
Added this entity and the circular buffer!